src/relay/transforms/simplify_expr.cc

```diff
@@ -146,7 +146,7 @@ class SimplifyConsecutiveCast : public DFPatternRewrite {
       // BFloat cast cannot be omitted
       return false;
     }
-    if (origin.code() < cast.code()) {
+    if (origin.code() < cast.code() && origin.bits() <= cast.bits()) {
```

Comment:
More general solution would be to calculate range of values dtypes represent. If moving to new dtype has range which is subset of old range it is a widen cast.

This perhaps is too harsh as it would block casting with int64 --> float32 or int32 --> float16 which may be relevant for quantization.

Reply:
That make sense. Shall we also block conversions like int64 -> uint32? So the condition will be only checking bits when the cast is between int/uint

Reply:
Hmm, Reading the comment of hierarchy int --> uint --> float does not make sense to me as uint can not represent negative numbers so it should be lowest on hiearchy in a sense.

I think this simplification pass is a bit hacky or at least it's assumptions are not super clear.  int --> uint hierarchy makes sense as long as we assume all uint casts are done on int types we know are only positive, or the cast clips negatives values to 0 and that is what we want. 

Not sure.

Reply:
If the point of this pass is making sure the output has the exact same output as old IRModule but with some casts remove, it actually doesn't do that job.

Reply:
But I think it's a bigger issue. I think PR is at least slightly better and safer than before so will approve. Will make a reminder to open issue here.